app/src/test/java/com/ably/tracking/demo/publisher/ably/LocationLoggerTest.kt

```diff
@@ -0,0 +1,131 @@
+    private val fileWriter = FakeLogFileWriter()
+
+    private val fileManager = FakeFileManager()
+
+    private val gson = Gson()
+
+    private val locationLogger = LocationLogger(fileWriter, fileManager, gson)
+
+    @Test
+    fun `on first call to logLocationUpdate file is created`() = runTest {
+        //given
+        val locationUpdate = prepareLocationUpdate(time = 1657782376167L)
+
+        //when
+        locationLogger.logLocationUpdate(locationUpdate)
+
+        //then
+        assertThat(fileWriter.fileName)
+            .isEqualTo("14.07_09:06:16_location.log")
+    }
+
+    @Test
+    fun `on call to logLocationUpdate location log is formatted and appended to the file`() =
+        runTest {
+            //given
+            val firstLocationUpdate = prepareLocationUpdate(time = 1657782376167L)
+            val secondLocationUpdate = prepareLocationUpdate(time = 1657782377167L)
+
+            //when
+            locationLogger.logLocationUpdate(firstLocationUpdate)
+            locationLogger.logLocationUpdate(secondLocationUpdate)
+
+            //then
+            assertThat(fileWriter.fileLines[0])
+                .isEqualTo("09:06:16 ${firstLocationUpdate.location}")
+            assertThat(fileWriter.fileLines[1])
+                .isEqualTo("09:06:17 ${secondLocationUpdate.location}")
+        }
+
+    @Test
+    fun `call to logLocationHistoryDataAndClose writes JSON-serialized data to a file`() = runTest {
+        //given
+        val locationUpdate = prepareLocationUpdate(time = 1657782376167L)
+        locationLogger.logLocationUpdate(locationUpdate)
+
+        val locationHistoryData = LocationHistoryData(
+            listOf(
+                GeoJsonMessage(
+                    type = "Feature",
+                    geometry = GeoJsonGeometry(
+                        type = "Point",
+                        coordinates = listOf(17.032827, 51.1052855, 0.0)
+                    ),
+                    properties = GeoJsonProperties(
+                        accuracyHorizontal = 699.999f,
+                        bearing = 100f,
+                        speed = 50.0f,
+                        time = 1231244123.0
+                    )
+                )
+            )
+        )
+        val locationHistoryDataJson =
+            "{" +
+                "\"events\":[" +
+                "{" +
+                "\"type\":\"Feature\"," +
+                "\"geometry\":{" +
+                "\"type\":\"Point\"," +
+                "\"coordinates\":[" +
+                "17.032827," +
+                "51.1052855," +
+                "0.0" +
+                "]" +
+                "}," +
+                "\"properties\":{" +
+                "\"accuracyHorizontal\":699.999," +
+                "\"bearing\":100.0," +
+                "\"speed\":50.0," +
+                "\"time\":1.231244123E9" +
+                "}" +
+                "}" +
+                "]," +
+                "\"version\":1" +
+                "}"
```

Comment:
I think you could use [Kotlin's multiline string syntax](https://kotlinlang.org/docs/java-to-kotlin-idioms-strings.html#use-multiline-strings) to make it cleaner :wink: 
```kotlin
val locationHistoryDataJson = """
  {
    "events" : [ ... ]
  }
""".trimIndent()
```

Reply:
On the note,  can I recommend using json file and read them from test classes. I find json embedded on source file very unreadable and unmaintainable.

Reply:
Unfortunately, this way line breaks are included in the string, while output from the actual serialization has no line breaks

Reply:
I gave you a wrong regex, for some reason this doesn't work
```kotlin
.replace(Regex.fromLiteral("\\s"), "")
```
but this one works :man_shrugging: 
```kotlin
.replace("\\s".toRegex(), "")
```

Reply:
https://github.com/ably/asset-tracking-android-rider-app-demo/pull/34/commits/60497c819b5894efce6deeaa99d56316329fb295